File: 04.Raspberry pi/2.Discovery/Test_Spi.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/spi/spidev.h>

#define SPI_DEVICE "/dev/spidev0.0"

void pabort(const char *s)
{
    perror(s);
    abort();
}

int main(void)
{
    int fd;
    uint8_t mode = SPI_MODE_0;
    uint8_t bits = 8;
    uint32_t speed = 50000;

    // Open SPI device
    if ((fd = open(SPI_DEVICE, O_RDWR)) < 0)
        pabort("can't open SPI device");

    // Set SPI mode
    if (ioctl(fd, SPI_IOC_WR_MODE, &mode) == -1)
        pabort("can't set SPI mode");

    // Set bits per word
    if (ioctl(fd, SPI_IOC_WR_BITS_PER_WORD, &bits) == -1)
        pabort("can't set bits per word");

    // Set max speed
    if (ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &speed) == -1)
        pabort("can't set max speed hz");

    // Data to be sent
    char data_to_send[] = "ismail is here bitchs";
    int length = strlen(data_to_send);

    // Prepare SPI transfer structure
    struct spi_ioc_transfer transfer = {
        .tx_buf = (unsigned long)data_to_send,
        .len = length,
        .speed_hz = speed,
        .bits_per_word = bits,
    };

    // Perform SPI transfer
    if (ioctl(fd, SPI_IOC_MESSAGE(1), &transfer) < 0)
        pabort("can't send spi message");


    // Close SPI device
    close(fd);

    return 0;
}

File: 04.Raspberry pi/2.Discovery/TestSpi.py
import spidev
import time

# Set SPI bus and device
spi_bus = 0
spi_device = 0

# Open SPI connection with mode 0
spi = spidev.SpiDev()
spi.open(spi_bus, spi_device)
spi.max_speed_hz = 50000  # You may need to adjust the speed based on your setup
spi.mode = 0  # Set SPI mode to 0 (CPOL=0, CPHA=0)

# Define the data to be sent
data_to_send = "niggas"

# Convert string to list of ASCII values
ascii_values = [ord(char) for char in data_to_send]

try:
    # Send data
    spi.xfer2(ascii_values)
    print("Data sent successfully")

finally:
    # Close SPI connection
    spi.close()

